### PR TITLE
Convert Diagnosticable to a mixin

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This converts `Diagnosticable` to be a mixin instead of an abstract class, so that it can be used to add diagnostics to classes which already have a base class.

It leaves in place the `DiagnosticableMixin` mixin, since there are some plugins that are still using it, and removing it would mean that those plugins wouldn't work with master branch Flutter anymore. `DiagnosticableMixin` will be removed once this mixin version of `Diagnosticable` makes its way to the stable branch.

## Related Issues

- https://github.com/flutter/flutter/issues/50498

## Tests

- No tests are needed, since we're just changing from a class to a mixin: the implementation stays the same.

## Breaking Change

- [X] No, this is *not* a breaking change, since no submitted customer tests use `Diagnosticable`.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 1]
```diff
@@ -3017,42 +3017,47 @@ class DiagnosticPropertiesBuilder {
...
-///  * [DiagnosticsProperty], which should be used to create leaf diagnostic
-///    nodes without properties or children. There are many
-///    [DiagnosticsProperty] subclasses to handle common use cases.
-abstract class Diagnosticable with DiagnosticableMixin {
-  /// Abstract const constructor. This constructor enables subclasses to provide
-  /// const constructors so that they can be used in const expressions.
-  const Diagnosticable();
+/// _This mixin is exists only to support plugins that require older Flutter
+/// versions: Use the identical mixin [Diagnosticable] instead for most code._
+mixin DiagnosticableMixin implements Diagnosticable {
```
Comment (darrenaustin) [inline packages/flutter/lib/src/foundation/diagnostics.dart]: Given that we intend to remove this, should it be marked as deprecated?

**Review by darrenaustin — APPROVED** [review]
Nice. Too bad we can't compose mixins, as that would have removed some duplication between Diagnosticable and DiagnosticableMixin.

Otherwise, it LGTM.

File: packages/flutter/lib/src/foundation/diagnostics.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/foundation/diagnostics.dart]: Yes, I definitely considered that, but mainly, I didn't want a plugin that wants to work with both master and stable to start throwing deprecation warnings on master: some projects would consider that an error and so they could no longer use that plugin on master, and the plugin author couldn't do anything about it until this change hit stable.  So I decided to wait to deprecate it until the change reached stable.

**Comment by HansMuller** [comment]
I'd starting reviewing this some time ago and then neglected to even submit what I'd written. So FWIW, now that it's too late :-)

**Comment by gspencergoog** [comment]
Did your writings uncover anything I should change?

**Comment by gspencergoog** [comment]
(Oh, I had to refresh GitHub to see them)

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 2]
```diff
@@ -3017,42 +3017,47 @@ class DiagnosticPropertiesBuilder {
...
-///    no children.
-///  * [DiagnosticsProperty], which should be used to create leaf diagnostic
-///    nodes without properties or children. There are many
-///    [DiagnosticsProperty] subclasses to handle common use cases.
-abstract class Diagnosticable with DiagnosticableMixin {
-  /// Abstract const constructor. This constructor enables subclasses to provide
-  /// const constructors so that they can be used in const expressions.
-  const Diagnosticable();
+/// _This mixin is exists only to support plugins that require older Flutter
+/// versions: Use the identical mixin [Diagnosticable] instead for most code._
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/foundation/diagnostics.dart]: That's no trailing `_`, that's the end of the italics for the paragraph.

**Comment by HansMuller** [comment]
The point about "most code" might be worth considering. I suppose youu mean all new code? https://github.com/flutter/flutter/pull/51495/files/baa502ed72e8ebdb12ba13d5687ee37d7d03f9c0#diff-e656bb511c2aa740e42637a30e61aabc

**Comment by gspencergoog** [comment]
The only code that needs to use that is plugin/package code that needs to work on stable. I'll clarify that, though.

**Comment by Zazo032** [comment]
@gspencergoog this is breaking builds for apps that use this plugin: https://github.com/Realank/flutter_datetime_picker

Is there any migration guide to update the plugin?

**Comment by gspencergoog** [comment]
Sorry about that.  There's no migration guide yet, but I'll make one today.

It's pretty straightforward, though:

If the plugin should be compatible with the `stable` channel and `master` channel, it should convert from this:
```dart
class MyClass extends Diagnosticable { ...
```
to this:
```dart
class MyClass with DiagnosticableMixin { ...
```
At a future time (once this change lands in `stable`), `DiagnosticableMixin` will be deprecated in favor of `Diagnosticable` and removed, however, so if the plugin only needs to be compatible with `master`, it can move directly to:

```dart
class MyClass with Diagnosticable { ...
```
and avoid the churn.

On the up side, this change also decreased code size by 1.5%...

**Comment by Zazo032** [comment]
No problem, I filled a PR to fix the plugin. Thanks for the size reduction! It was 0.2mb on our appbundle 🔥

**Pull request merged**